Fix table function execution without partitioning (v2)

findepi opened the pull request:
Previously, when table function did not declare partitioning it would run single-threaded and first buffer all data in memory, like a one big WINDOW. After the change, the local execution processes input pages in a streaming fashion.

Fixes https://github.com/trinodb/trino/issues/20398

Alternative to https://github.com/trinodb/trino/pull/21378, with different `TableFunctionDataProcessor` lifecycle. The implementation creates one TableFunctionDataProcessor per operator for streaming processing.

findepi committed:
Fix table function execution without partitioning

Previously, when table function did not declare partitioning, it would
be globally distributed, but on a worker node it would run
single-threaded and first buffer all data in memory, like a one big
WINDOW. After the change, the local execution processes input pages in a
streaming fashion.

This commit also fixes property derivations for a case where table
function is partitioned on empty list of symbols (global grouping).

findepi committed:
Rename interface to accomodate non-partitioned data

findepi commented:
failure is related.

hovaesco commented:
I've tested the fix and it helps with https://github.com/trinodb/trino/issues/20398 however when writing the data using table function, it still does not distribute the data evenly and there is no way to control the number of files being written.

findepi commented:
( rebased to resolve conflicts, no other changes )

findepi commented:
Planned changes moved out to https://github.com/trinodb/trino/pull/21710

findepi committed:
Run tests with more parallelism

During migration from TestNG to JUnit `@TestInstance(PER_CLASS)`
annotation was added, but it implies single-threaded execution. Restore
previous parallelism: either add `@Execution(CONCURRENT)` or inherit it
from base class.

tbaeg commented on core/trino-main/src/main/java/io/trino/operator/function/StreamTableFunctionInput.java:
```diff
@@ -0,0 +1,125 @@
+                    pagesIndex,
+                    0,
+                    pagesIndex.getPositionCount(),
+                    new TableFunctionDataProcessor()
```
Do we need to wrap in a new anonymous class?

Locally, I directly passing the `tableFunction` which also resolved the `TestJsonTable` failures for me.

tbaeg commented:
Wanted to see if there was any movement on this. Happy to assist with this as well.
